src/cpp/session/modules/SessionCopilot.cpp

```diff
@@ -0,0 +1,894 @@
+               std::back_inserter(keys));
+
+      for (auto&& key : keys)
+      {
+         auto&& continuation = s_pendingContinuations[key];
+         auto elapsedTime = currentTime - continuation.time();
+         if (elapsedTime.seconds() > 10)
+         {
+            DLOG("Cancelling old contiuation with id '{}'.", key);
+            continuation.cancel();
+            s_pendingContinuations.erase(key);
+         }
+      }
+   }
+
+   // process any pending requests
+   while (!s_pendingResponses.empty())
+   {
+      // Try to parse this as JSON, and see if we received the response we expect.
+      // The agent might be chatty inbetween a request and a response, so we can
+      // drop any output not specifically for us.
+      std::string response = s_pendingResponses.front();
+      s_pendingResponses.pop();
+
+      json::Object responseJson;
+      Error error = responseJson.parse(response);
+      if (error)
+      {
+         LOG_ERROR(error);
+         continue;
+      }
+
+      // Check if this is a 'LogMessage' response. Should we log these in verbose mode?
+      json::Value methodJson = responseJson["method"];
+      if (methodJson.isString())
+      {
+         std::string method = methodJson.getString();
+         if (method == "LogMessage")
+            continue;
+      }
+
+      // Check the response id. This will be missing for notifications; we may receive
+      // a flurry of progress notifications when requesting completions from Copilot.
+      // We might want to handle these somehow. Perhaps hook them up to some status widget?
+      json::Value requestIdJson = responseJson["id"];
+      if (!requestIdJson.isString())
+      {
+         DLOG("Ignoring response with missing id.");
+         continue;
+      }
+
+      // Check for a continuation handler, and invoke it if available.
+      std::string requestId = requestIdJson.getString();
+      if (s_pendingContinuations.count(requestId))
+      {
+         DLOG("Invoking continuation with id '{}'.", requestId);
+         s_pendingContinuations[requestId].invoke(responseJson);
+         s_pendingContinuations.erase(requestId);
+      }
+   }
+}
+
+void onPreferencesSaved()
+{
+   // Check for preference change
+   bool oldEnabled = s_copilotEnabled;
+   bool newEnabled = prefs::userPrefs().copilotEnabled();
+   if (oldEnabled == newEnabled)
+      return;
+
+   // Update our preference
+   s_copilotEnabled = newEnabled;
+
+   // Start or stop the agent as appropriate
+   if (s_copilotEnabled)
+   {
+      startAgent();
+   }
+   else
+   {
+      stopAgent();
+   }
+}
+
+void onUserPrefsChanged(const std::string& layer,
+                        const std::string& name)
+{
+   if (name == "copilot_enabled")
+   {
+      onPreferencesSaved();
+   }
+}
+
+// TODO: Does this need to be deferred? We might miss out
+// on some open documents if events are fired before
+// deferred init has finished.
+void onDeferredInit(bool newSession)
```

Comment:
We should try to move this out of deferred init, or check if the client init event gives us the necessary source document information.